### PR TITLE
DiligentCore.NET: Base implementation

### PR DESCRIPTION
**Pull request by MikhailGorobets**

### CODE REVIEW

File: Graphics/GraphicsEngine.NET/BlendState.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using SharpGen.Runtime;
+
+namespace Diligent.Core;
+
+public partial struct RenderTargetBlendDesc
```
Comment (TheMostDiligent) [inline Graphics/GraphicsEngine.NET/BlendState.cs:35]: What can't be auto-generated here?

File: Graphics/GraphicsEngine.NET/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+if(D3D11_SUPPORTED)
+    add_dependencies(Diligent-GraphicsEngine.NET Diligent-GraphicsEngineD3D11-shared)
+endif()
+
+if(D3D12_SUPPORTED)
+    add_dependencies(Diligent-GraphicsEngine.NET Diligent-GraphicsEngineD3D12-shared)
+endif()
+
+if(VULKAn_SUPPORTED)
```
Comment (TheMostDiligent) [inline Graphics/GraphicsEngine.NET/CMakeLists.txt:14]: Typo (`VULKAn`)

File: Graphics/GraphicsEngine.NET/DeviceContext.cs [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+{
+    public DrawAttribs()
+    {
+        NumVertices = default;
+        Flags = DrawFlags.None;
+        NumInstances = 1;
+        StartVertexLocation = default;
+        FirstInstanceLocation = default;
+    }
+}
```
Comment (TheMostDiligent) [inline Graphics/GraphicsEngine.NET/DeviceContext.cs:45]: Here and everywhere else: why constructors like this one are not auto-generated?

File: Graphics/GraphicsEngine.NET/Diligent-GraphicsEngine.NET.csproj [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+		<PackageProjectUrl>https://github.com/DiligentGraphics/DiligentCore</PackageProjectUrl>
+		<RepositoryUrl>https://github.com/DiligentGraphics/DiligentCore</RepositoryUrl>
+		<RepositoryType>git</RepositoryType>
+		<Company>Diligent Graphics LLC</Company>
+	</PropertyGroup>
+
+	<ItemGroup>
+		<SharpGenMapping Include="Mapping.xml" />
+	</ItemGroup>
+</Project>
```
Comment (TheMostDiligent) [inline Graphics/GraphicsEngine.NET/Diligent-GraphicsEngine.NET.csproj:19]: Can this be generated by CMake?
Cant this be built not my Visual Studio?

File: Graphics/GraphicsEngine.NET/Mapping.xml [hunk 5]
```diff
@@ -0,0 +1,443 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (TheMostDiligent) [inline Graphics/GraphicsEngine.NET/Mapping.xml:1]: Can this file be auto-generated?

File: Graphics/GraphicsEngine/interface/GraphicsTypes.h [hunk 6]
```diff
@@ -1895,12 +1895,20 @@ struct Version
...
 
+#if !defined(DILIGENT_SHARP_GEN)
     template <typename T, typename = typename std::enable_if<std::is_integral<T>::value>::type>
     constexpr Version(T _Major, T _Minor) noexcept :
         Major{static_cast<decltype(Major)>(_Major)},
         Minor{static_cast<decltype(Minor)>(_Minor)}
     {
     }
+#else
+    constexpr Version(Uint32 _Major, Uint32 _Minor) noexcept :
```
Comment (TheMostDiligent) [inline Graphics/GraphicsEngine/interface/GraphicsTypes.h:1906]: Is template the problem here?

File: Tests/DiligentCoreTest.NET/Program.cs [hunk 7]
```diff
@@ -0,0 +1,316 @@
+/*
+ *  Copyright 2019-2023 Diligent Graphics LLC
```
Comment (TheMostDiligent) [inline Tests/DiligentCoreTest.NET/Program.cs:2]: Is this a test program?

**Comment by MrScautHD** [comment]
Get Diligent Engine a C# update too?

File: Graphics/GraphicsEngine/interface/GraphicsTypes.h (hunk 6 above)
Comment (MikhailGorobets) [reply Graphics/GraphicsEngine/interface/GraphicsTypes.h:1913]: CastXML uses clang 13.0.0 for parsing C++ code. The latest C++ headers in MSVC require clang version 15.0.0
https://github.com/CastXML/CastXMLSuperbuild/issues/53

**Pull request closed without merging**